# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/style/FontFace.h
# Commit: 93b7e171c74a
# Full Hash: 93b7e171c74a6aa88d98979009eb0e299e2b61a3
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2022-06-29 03:44:30
# Regressor Bug: 1771493
# File Overlap Count: 1
# Description:
#   Bug 1771493 - Part 1. Split FontFace(Set) into DOM facing and implementation classes. r=emilio
#   
#   This patch aims to split the implementation details of FontFace(Set)
#   into FontFace(Set)Impl classes. The Impl variants have threadsafe
#   refcounting and are intended to be the basis for worker support for
# ==============================================================================

diff -r c45dacd63805 -r 93b7e171c74a layout/style/FontFace.h
--- a/layout/style/FontFace.h	Tue Jun 28 14:12:34 2022 +0000
+++ b/layout/style/FontFace.h	Tue Jun 28 14:27:20 2022 +0000
@@ -26,7 +26,9 @@
 class CSSFontFaceRule;
 class FontFaceBufferSource;
 struct FontFaceDescriptors;
+class FontFaceImpl;
 class FontFaceSet;
+class FontFaceSetImpl;
 class Promise;
 class UTF8StringOrArrayBufferOrArrayBufferView;
 }  // namespace dom
@@ -36,41 +38,8 @@
 
 class FontFace final : public nsISupports, public nsWrapperCache {
   friend class mozilla::PostTraversalTask;
-  friend class FontFaceBufferSource;
-  friend class Entry;
 
  public:
-  class Entry final : public gfxUserFontEntry {
-    friend class FontFace;
-
-   public:
-    Entry(gfxUserFontSet* aFontSet,
-          const nsTArray<gfxFontFaceSrc>& aFontFaceSrcList, WeightRange aWeight,
-          StretchRange aStretch, SlantStyleRange aStyle,
-          const nsTArray<gfxFontFeature>& aFeatureSettings,
-          const nsTArray<gfxFontVariation>& aVariationSettings,
-          uint32_t aLanguageOverride, gfxCharacterMap* aUnicodeRanges,
-          StyleFontDisplay aFontDisplay, RangeFlags aRangeFlags,
-          float aAscentOverride, float aDescentOverride, float aLineGapOverride,
-          float aSizeAdjust)
-        : gfxUserFontEntry(aFontSet, aFontFaceSrcList, aWeight, aStretch,
-                           aStyle, aFeatureSettings, aVariationSettings,
-                           aLanguageOverride, aUnicodeRanges, aFontDisplay,
-                           aRangeFlags, aAscentOverride, aDescentOverride,
-                           aLineGapOverride, aSizeAdjust) {}
-
-    virtual void SetLoadState(UserFontLoadState aLoadState) override;
-    virtual void GetUserFontSets(nsTArray<gfxUserFontSet*>& aResult) override;
-    const AutoTArray<FontFace*, 1>& GetFontFaces() { return mFontFaces; }
-
-   protected:
-    // The FontFace objects that use this user font entry.  We need to store
-    // an array of these, not just a single pointer, since the user font
-    // cache can return the same entry for different FontFaces that have
-    // the same descriptor values and come from the same origin.
-    AutoTArray<FontFace*, 1> mFontFaces;
-  };
-
   NS_DECL_CYCLE_COLLECTING_ISUPPORTS
   NS_DECL_CYCLE_COLLECTION_SCRIPT_HOLDER_CLASS(FontFace)
 
@@ -82,77 +51,6 @@
                                                   FontFaceSet* aFontFaceSet,
                                                   RawServoFontFaceRule* aRule);
 
-  RawServoFontFaceRule* GetRule() { return mRule; }
-
-  bool HasLocalSrc() const;
-  Maybe<StyleComputedFontWeightRange> GetFontWeight() const;
-  Maybe<StyleComputedFontStretchRange> GetFontStretch() const;
-  Maybe<StyleComputedFontStyleDescriptor> GetFontStyle() const;
-  Maybe<StyleFontDisplay> GetFontDisplay() const;
-  void GetFontFeatureSettings(nsTArray<gfxFontFeature>&) const;
-  void GetFontVariationSettings(nsTArray<gfxFontVariation>&) const;
-  void GetSources(nsTArray<StyleFontFaceSourceListComponent>&) const;
-  Maybe<StyleFontLanguageOverride> GetFontLanguageOverride() const;
-  Maybe<StylePercentage> GetAscentOverride() const;
-  Maybe<StylePercentage> GetDescentOverride() const;
-  Maybe<StylePercentage> GetLineGapOverride() const;
-  Maybe<StylePercentage> GetSizeAdjust() const;
-
-  gfxUserFontEntry* CreateUserFontEntry();
-  gfxUserFontEntry* GetUserFontEntry() const { return mUserFontEntry; }
-  void SetUserFontEntry(gfxUserFontEntry* aEntry);
-
-  /**
-   * Returns whether this object is in the specified FontFaceSet.
-   */
-  bool IsInFontFaceSet(FontFaceSet* aFontFaceSet) const;
-
-  void AddFontFaceSet(FontFaceSet* aFontFaceSet);
-  void RemoveFontFaceSet(FontFaceSet* aFontFaceSet);
-
-  FontFaceSet* GetPrimaryFontFaceSet() const { return mFontFaceSet; }
-
-  /**
-   * Gets the family name of the FontFace as a raw string (such as 'Times', as
-   * opposed to GetFamily, which returns a CSS-escaped string, such as
-   * '"Times"').  Returns null if a valid family name was not available.
-   */
-  nsAtom* GetFamilyName() const;
-
-  /**
-   * Returns whether this object is CSS-connected, i.e. reflecting an
-   * @font-face rule.
-   */
-  bool HasRule() const { return mRule; }
-
-  /**
-   * Breaks the connection between this FontFace and its @font-face rule.
-   */
-  void DisconnectFromRule();
-
-  /**
-   * Returns whether there is an ArrayBuffer or ArrayBufferView of font
-   * data.
-   */
-  bool HasFontData() const;
-
-  /**
-   * Creates a gfxFontFaceBufferSource to represent the font data
-   * in this object.
-   */
-  already_AddRefed<gfxFontFaceBufferSource> CreateBufferSource();
-
-  /**
-   * Gets a pointer to and the length of the font data stored in the
-   * ArrayBuffer or ArrayBufferView.
-   */
-  bool GetData(uint8_t*& aBuffer, uint32_t& aLength);
-
-  /**
-   * Returns the value of the unicode-range descriptor as a gfxCharacterMap.
-   */
-  gfxCharacterMap* GetUnicodeRangeAsCharacterMap();
-
   // Web IDL
   static already_AddRefed<FontFace> Constructor(
       const GlobalObject& aGlobal, const nsACString& aFamily,
@@ -190,65 +88,30 @@
   Promise* Load(ErrorResult& aRv);
   Promise* GetLoaded(ErrorResult& aRv);
 
- private:
-  FontFace(nsISupports* aParent, FontFaceSet* aFontFaceSet);
-  ~FontFace();
-
-  void InitializeSource(const UTF8StringOrArrayBufferOrArrayBufferView&);
-
-  // Helper function for Load.
-  void DoLoad();
-
-  // Helper function for the descriptor setter methods.
-  // Returns true if the descriptor was modified, false if descriptor is
-  // unchanged (which may not be an error: check aRv for actual failure).
-  bool SetDescriptor(nsCSSFontDesc aFontDesc, const nsACString& aValue,
-                     ErrorResult& aRv);
+  FontFaceImpl* GetImpl() const { return mImpl; }
 
-  /**
-   * Sets all of the descriptor values in mDescriptors using values passed
-   * to the JS constructor.
-   * Returns true on success, false if parsing any descriptor failed.
-   */
-  bool SetDescriptors(const nsACString& aFamily,
-                      const FontFaceDescriptors& aDescriptors);
-
-  /**
-   * Called when a descriptor has been modified, so font-face sets can
-   * be told to refresh.
-   */
-  void DescriptorUpdated();
-
-  /**
-   * Sets the current loading status.
-   */
-  void SetStatus(FontFaceLoadStatus aStatus);
-
-  void GetDesc(nsCSSFontDesc aDescID, nsACString& aResult) const;
+  void Destroy();
+  void MaybeResolve();
+  void MaybeReject(nsresult aResult);
 
   already_AddRefed<URLExtraData> GetURLExtraData() const;
 
-  RawServoFontFaceRule* GetData() const {
-    return HasRule() ? mRule : mDescriptors;
-  }
+ private:
+  explicit FontFace(nsISupports* aParent);
+  ~FontFace();
 
   /**
    * Returns and takes ownership of the buffer storing the font data.
    */
   void TakeBuffer(uint8_t*& aBuffer, uint32_t& aLength);
 
-  // Acts like mLoaded->MaybeReject(aResult), except it doesn't create mLoaded
-  // if it doesn't already exist.
-  void Reject(nsresult aResult);
-
   // Creates mLoaded if it doesn't already exist. It may immediately resolve or
   // reject mLoaded based on mStatus and mLoadedRejection.
   void EnsurePromise();
 
-  void DoResolve();
-  void DoReject(nsresult aResult);
+  nsCOMPtr<nsISupports> mParent;
 
-  nsCOMPtr<nsISupports> mParent;
+  RefPtr<FontFaceImpl> mImpl;
 
   // A Promise that is fulfilled once the font represented by this FontFace is
   // loaded, and is rejected if the load fails. This promise is created lazily
@@ -257,62 +120,6 @@
 
   // Saves the rejection code for mLoaded if mLoaded hasn't been created yet.
   nsresult mLoadedRejection;
-
-  // The @font-face rule this FontFace object is reflecting, if it is a
-  // rule backed FontFace.
-  RefPtr<RawServoFontFaceRule> mRule;
-
-  // The FontFace object's user font entry.  This is initially null, but is set
-  // during FontFaceSet::UpdateRules and when a FontFace is explicitly loaded.
-  RefPtr<Entry> mUserFontEntry;
-
-  // The current load status of the font represented by this FontFace.
-  // Note that we can't just reflect the value of the gfxUserFontEntry's
-  // status, since the spec sometimes requires us to go through the event
-  // loop before updating the status, rather than doing it immediately.
-  FontFaceLoadStatus mStatus;
-
-  // Represents where a FontFace's data is coming from.
-  enum SourceType {
-    eSourceType_FontFaceRule = 1,
-    eSourceType_URLs,
-    eSourceType_Buffer
-  };
-
-  // Where the font data for this FontFace is coming from.
-  SourceType mSourceType;
-
-  // If the FontFace was constructed with an ArrayBuffer(View), this is a
-  // copy of the data from it.
-  uint8_t* mSourceBuffer;
-  uint32_t mSourceBufferLength;
-
-  // The values corresponding to the font face descriptors, if we are not
-  // a rule backed FontFace object.  For rule backed objects, we use
-  // the descriptors stored in mRule.
-  // FIXME This should hold a unique ptr to just the descriptors inside,
-  // so that we don't need to create a rule for it and don't need to
-  // assign a fake line number and column number. See bug 1450904.
-  RefPtr<RawServoFontFaceRule> mDescriptors;
-
-  // The value of the unicode-range descriptor as a gfxCharacterMap.  Valid
-  // only when mUnicodeRangeDirty is false.
-  RefPtr<gfxCharacterMap> mUnicodeRange;
-
-  // The primary FontFaceSet this FontFace is associated with,
-  // regardless of whether it is currently "in" the set.
-  RefPtr<FontFaceSet> mFontFaceSet;
-
-  // Other FontFaceSets (apart from mFontFaceSet) that this FontFace
-  // appears in.
-  nsTArray<RefPtr<FontFaceSet>> mOtherFontFaceSets;
-
-  // Whether mUnicodeRange needs to be rebuilt before being returned from
-  // GetUnicodeRangeAsCharacterMap.
-  bool mUnicodeRangeDirty;
-
-  // Whether this FontFace appears in mFontFaceSet.
-  bool mInFontFaceSet;
 };
 
 }  // namespace mozilla::dom